zcash_client_backend/src/data_api.rs

```diff
@@ -0,0 +1,583 @@
+
+    /// Returns the block hash for the block at the maximum height known
+    /// in stored data.
+    fn get_max_height_hash(&self) -> Result<Option<(BlockHeight, BlockHash)>, Self::Error> {
+        self.block_height_extrema()
+            .and_then(|extrema_opt| {
+                extrema_opt
+                    .map(|(_, max_height)| {
+                        self.get_block_hash(max_height)
+                            .map(|hash_opt| hash_opt.map(move |hash| (max_height, hash)))
+                    })
+                    .transpose()
+            })
+            .map(|oo| oo.flatten())
+    }
+
+    /// Returns the block height in which the specified transaction was mined.
+    fn get_tx_height(&self, txid: TxId) -> Result<Option<BlockHeight>, Self::Error>;
+
+    /// Returns the payment address for the specified account, if the account
+    /// identifier specified refers to a valid account for this wallet.
+    fn get_address(&self, account: AccountId) -> Result<Option<PaymentAddress>, Self::Error>;
+
+    /// Returns all extended full viewing keys known about by this wallet
+    fn get_extended_full_viewing_keys(
+        &self,
+    ) -> Result<HashMap<AccountId, ExtendedFullViewingKey>, Self::Error>;
+
+    /// Checks whether the specified extended full viewing key is
+    /// associated with the account.
+    fn is_valid_account_extfvk(
+        &self,
+        account: AccountId,
+        extfvk: &ExtendedFullViewingKey,
+    ) -> Result<bool, Self::Error>;
+
+    /// Returns the wallet balance for the specified account.
+    ///
+    /// This balance amount is the raw balance of all transactions in known
+    /// mined blocks, irrespective of confirmation depth. A default implementation
+    /// is provided, but implementations of this trait may override the default
+    /// implementation for efficiency.
+    fn get_balance(&self, account: AccountId) -> Result<Amount, Self::Error> {
+        if let Some((_, tip)) = self.block_height_extrema()? {
+            self.get_verified_balance(account, tip)
+        } else {
+            Ok(Amount::zero())
+        }
+    }
+
+    /// Returns the wallet balance for an account as of the specified block
+    /// height. and
+    ///
+    /// This may be used to obtain a balance that ignores notes that have been
+    /// received so recently that they are not yet deemed spendable.
+    fn get_verified_balance(
+        &self,
+        account: AccountId,
+        anchor_height: BlockHeight,
+    ) -> Result<Amount, Self::Error>;
+
+    /// Returns the memo for a received note, if it is known and a valid UTF-8 string.
+    fn get_received_memo_as_utf8(
+        &self,
+        id_note: Self::NoteRef,
+    ) -> Result<Option<String>, Self::Error>;
+
+    /// Returns the memo for a sent note, if it is known and a valid UTF-8 string.
+    fn get_sent_memo_as_utf8(&self, id_note: Self::NoteRef) -> Result<Option<String>, Self::Error>;
```

Comment:
Why do these need to be separate? What happens if you call `get_received_memo_as_utf8` on a note you've sent, or vice versa?

Also, why restrict the API to only working on memos that are valid UTF-8?

Reply:
The IDs don't line up (they are two separate tables in SQLite because they store different information).

Reply:
Maybe the right thing here would be to have a single `NoteRef` type, and a single `get_memo_as_utf8` but then have the sqlite backend specialize that `NoteRef` type to an enum type, which is then used to dispatch to the correct table. I'll make that change.

Reply:
https://github.com/nuttycom/librustzcash/pull/3

I'm adding this as a PR to the base branch so that it can be reviewed independently. @str4d does this make sense?

Reply:
That resolved the point about separate APIs. Filed #331 for this question:

> why restrict the API to only working on memos that are valid UTF-8?